instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java

```diff
@@ -53,33 +53,22 @@ public static void onEnter(
     }
 
     Context attachedContext = tracer().getServerContext(httpServletRequest);
-    if (attachedContext != null) {
+    if (attachedContext != null && tracer().needsRescoping(attachedContext)) {
       // We are inside nested servlet/filter/app-server span, don't create new span
-      if (tracer().needsRescoping(attachedContext)) {
-        attachedContext =
-            tracer().updateContext(attachedContext, httpServletRequest, mappingResolver, servlet);
-        scope = attachedContext.makeCurrent();
-        return;
-      }
-
-      // We already have attached context to request but this could have been done by app server
-      // instrumentation, if needed update span with info from current request.
-      Context currentContext = Java8BytecodeBridge.currentContext();
-      Context updatedContext =
-          tracer().updateContext(currentContext, httpServletRequest, mappingResolver, servlet);
-      if (updatedContext != currentContext) {
-        // runOnceUnderAppServer updated context, need to re-scope
-        scope = updatedContext.makeCurrent();
-      }
+      attachedContext =
+          tracer().updateContext(attachedContext, httpServletRequest, mappingResolver, servlet);
+      scope = attachedContext.makeCurrent();
       return;
     }
 
     Context currentContext = Java8BytecodeBridge.currentContext();
     if (currentContext != null
```

Comment:
By the way, `currentContext()` can never return null - you can simplify that bit too 😄

Reply:
Thanks, removed it.